api/proto/index/v2/index.proto

```diff
@@ -0,0 +1,161 @@
+syntax = "proto3";
+
+import "google/protobuf/timestamp.proto";
+import "google/protobuf/empty.proto";
+
+package index.v2;
+option go_package = "./index/v2;index";
+
+// Service provides an external API for managing the index records
+service Service {
+  // Create allows to new records for a search.
+  rpc Create(CreateRecordsRequest) returns (CreateRecordsResult);
+  // Create allows to create a new records, streaming the records data in the request
+  rpc CreateWithStreamData(stream CreateIndexStreamRequest) returns (CreateRecordsResult);
+  // DeleteNode allows to delete a Node and all its children identified by the path provided
+  rpc DeleteNode(Path) returns (google.protobuf.Empty);
+  // ListNodes returns all known children for the Path provided
+  rpc ListNodes(Path) returns (Nodes);
+  // Patch allows to insert, update or delete an index's records
+  rpc PatchRecords(PatchRecordsRequest) returns (PatchRecordsResult);
+  // ListRecords returns list of records for a path associated with it
+  rpc ListRecords(ListRequest) returns (ListRecordsResult);
+  // Search runs the search across all indexes by the specified query. Result will
+  // be ordered by the ranks for the request
+  rpc Search(SearchRecordsRequest) returns (SearchRecordsResult);
+}
+
+// Path structure describes the list of nodes which identify a final node in the object tree.
+message Path {
+  // path is the nodes names separated by '/' like in the computer filesystem
+  string path = 1;
+}
+
+enum NodeType {
+  FOLDER = 0;
+  DOCUMENT = 1;
+}
+
+message Node {
```

Comment:
Is it allowed to have records of different formats in a single Node? If no, how to detect that the vector of new records is allowed for the Node (checking already existing records)?

Reply:
How to deal with different versions of the same format? In particular, should different versions of the same format be allowed in the same Node? (the simplest, is not to allow and force reindex)

Reply:
> Is it allowed to have records of different formats in a single Node? If no, how to detect that the vector of new records is allowed for the Node (checking already existing records)?

yes, it is allowed. With a node multiple records with different formats may be associated. For example for a pdf document we may have:
- title - metadata record
- the filename - metadata record
- pdf - the content records
...

> How to deal with different versions of the same format? In particular, should different versions of the same format be allowed in the same Node? (the simplest, is not to allow and force reindex)

Each record has it's unique ID within the node. So we can change/delete it

Reply:
Ok, basically if we support different formats, then a version of the same format is just a new format.